// @hex-engine/2d/dist/Components/AnimationSheet.js
import { useNewComponent, useType } from "@hex-engine/core";
import SpriteSheet from "./SpriteSheet";

/**
 * A Component representing an AnimationSheet image; that is, a filmstrip-style image
 * of sprites which should be rendered in a particular sequence as part of an animation.
 */
export default function AnimationSheet(_ref) {
  var url = _ref.url,
      tileWidth = _ref.tileWidth,
      tileHeight = _ref.tileHeight,
      animations = _ref.animations;
  useType(AnimationSheet);
  var spriteSheet = useNewComponent(function () {
    return SpriteSheet({
      url: url,
      tileWidth: tileWidth,
      tileHeight: tileHeight
    });
  });
  var currentAnim = animations["default"];
  return {
    /** The current animation, that frames will be drawn from. */
    get currentAnim() {
      return currentAnim;
    },

    /** Sets which animation is the current animation. */
    set currentAnim(nextValue) {
      currentAnim = nextValue;
    },

    /** Draws the current animation's current frame into the provided canvas context. */
    draw: function draw(context) {
      var _ref2 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
          x = _ref2.x,
          y = _ref2.y,
          width = _ref2.width,
          height = _ref2.height;

      var frame = currentAnim.currentFrame;
      spriteSheet.draw(context, {
        x: x,
        y: y,
        width: width,
        height: height,
        tileIndex: frame.data
      });
    }
  };
}
