// @hex-engine/inspector/dist/PausedOverlay.js
import React from "react";
import Button from "./Button";
import { ResumeIcon } from "./Icons";
export default function PausedOverlay(_ref) {
  var runLoop = _ref.runLoop;
  return React.createElement("div", {
    style: {
      position: "fixed",
      top: "calc(50vh - 64px)",
      left: "calc(50vw - 64px)",
      display: "flex",
      alignItems: "center",
      justifyContent: "center"
    }
  }, React.createElement("div", {
    style: {
      width: "128px",
      height: "128px",
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
      boxShadow: "0px 0px 10px rgba(0, 0, 0, 0.5)",
      backgroundColor: "rgba(238, 238, 238)",
      borderRadius: "16px"
    }
  }, React.createElement(Button, {
    onClick: function onClick() {
      localStorage.inspectorPauseOnStart = "false";
      runLoop.resume();
    },
    title: "Resume execution",
    style: {
      color: "#008eff"
    }
  }, React.createElement(ResumeIcon, {
    width: 128,
    height: 128
  }))));
}
