// hex-engine-scripts/dist/commands/dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = dev;

var _chalk = _interopRequireDefault(require("chalk"));

var _webpack = _interopRequireDefault(require("webpack"));

var _webpackDevServer = _interopRequireDefault(require("webpack-dev-server"));

var _clearConsole = _interopRequireDefault(require("react-dev-utils/clearConsole"));

var _WebpackDevServerUtils = require("react-dev-utils/WebpackDevServerUtils");

var _makeWebpackConfig = _interopRequireDefault(require("../makeWebpackConfig"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function dev(_x) {
  return _dev.apply(this, arguments);
}

function _dev() {
  _dev = _asyncToGenerator(
  /*#__PURE__*/
  regeneratorRuntime.mark(function _callee(options) {
    var isInteractive, port, urls, webpackConfig, devSocket, compiler, devServer;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            process.env.BABEL_ENV = "development";
            process.env.NODE_ENV = "development";
            isInteractive = process.stdout.isTTY;
            _context.next = 5;
            return (0, _WebpackDevServerUtils.choosePort)("localhost", options.port || 8080);

          case 5:
            port = _context.sent;

            if (port) {
              _context.next = 8;
              break;
            }

            return _context.abrupt("return");

          case 8:
            urls = (0, _WebpackDevServerUtils.prepareUrls)("http", "localhost", port);
            webpackConfig = options.lib ? (0, _makeWebpackConfig["default"])({
              mode: "development",
              srcFile: "lib",
              outDir: "lib",
              library: options.lib,
              title: options.title
            }) : (0, _makeWebpackConfig["default"])({
              mode: "development",
              srcFile: "index",
              outDir: "dist",
              title: options.title
            });
            devSocket = {
              warnings: function warnings(_warnings) {
                return (// @ts-ignore
                  devServer.sockWrite(devServer.sockets, "warnings", _warnings)
                );
              },
              errors: function errors(_errors) {
                return (// @ts-ignore
                  devServer.sockWrite(devServer.sockets, "errors", _errors)
                );
              }
            }; // Create a webpack compiler that is configured with custom messages.
            // @ts-ignore

            compiler = (0, _WebpackDevServerUtils.createCompiler)({
              appName: options.title || "hex-engine game",
              // @ts-ignore
              config: webpackConfig,
              devSocket: devSocket,
              urls: urls,
              useYarn: true,
              useTypeScript: true,
              tscCompileOnError: false,
              webpack: _webpack["default"]
            });
            devServer = new _webpackDevServer["default"](compiler, {
              compress: true,
              clientLogLevel: "none",
              hot: true,
              publicPath: "/",
              quiet: true,
              overlay: true,
              historyApiFallback: {
                // Paths with dots should still use the history fallback.
                // See https://github.com/facebook/create-react-app/issues/387.
                disableDotRule: true
              }
            });
            return _context.abrupt("return", new Promise(function (resolve, reject) {
              devServer.listen(port, function (err) {
                if (err) {
                  reject(err);
                }

                if (isInteractive) {
                  (0, _clearConsole["default"])();
                }

                console.log(_chalk["default"].cyan("Starting the development server...\n"));
                process.on("SIGINT", function () {
                  devServer.close();
                  resolve();
                });
                process.on("SIGTERM", function () {
                  devServer.close();
                  resolve();
                });
              });
            }));

          case 14:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _dev.apply(this, arguments);
}
