// kame/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var bundler_1 = __importDefault(require("./bundler"));
var runtime_1 = __importDefault(require("./runtime"));
var config_1 = require("./config");
var default_loader_1 = __importDefault(require("./default-loader"));
exports.defaultLoader = default_loader_1.default;
var default_resolver_1 = require("./default-resolver");
exports.defaultResolver = default_resolver_1.resolve;
var default_runtime_eval_1 = __importDefault(require("./default-runtime-eval"));
exports.defaultRuntimeEval = default_runtime_eval_1.default;
var default_instance_1 = require("./default-instance");
exports.Runtime = default_instance_1.Runtime;
exports.Bundler = default_instance_1.Bundler;
function configure(inputConfig) {
    if (inputConfig === void 0) { inputConfig = {}; }
    var config = config_1.readConfig(inputConfig);
    var Bundler = bundler_1.default(config);
    var Runtime = runtime_1.default(config);
    return {
        Bundler: Bundler,
        Runtime: Runtime,
    };
}
exports.configure = configure;
