// @test-it/core/dist/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var util_1 = __importDefault(require("util"));
var default_reporter_1 = __importDefault(require("@test-it/default-reporter"));
var default_loader_1 = __importDefault(require("@test-it/default-loader"));
var debug_1 = __importDefault(require("debug"));
var debug = debug_1.default("@test-it/core:config.ts");
function normalizeConfig(config) {
    var _a;
    debug("Parsing PartialConfig: " + util_1.default.inspect(config));
    return {
        testFiles: config.testFiles,
        reporters: config.reporters && config.reporters.length > 0
            ? config.reporters
            : [new default_reporter_1.default()],
        loader: (_a = config.loader) !== null && _a !== void 0 ? _a : default_loader_1.default,
        resolver: config.resolver,
        seed: config.seed,
        updateSnapshots: config.updateSnapshots,
        testSetupFiles: config.testSetupFiles || [],
        shouldAbort: config.shouldAbort || (function () { return false; }),
    };
}
exports.normalizeConfig = normalizeConfig;
