// @test-it/core/dist/test-name-reporter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var util_1 = __importDefault(require("util"));
var debug_1 = __importDefault(require("debug"));
var debug = debug_1.default("@test-it/core:test-name-reporter.ts");
var TestNameReporter = /** @class */ (function () {
    function TestNameReporter() {
        this.stack = [];
    }
    TestNameReporter.prototype.getCurrentTestFile = function () {
        return this.stack[0];
    };
    TestNameReporter.prototype.getCurrentTestName = function () {
        return this.stack.slice(1, this.stack.length).join(" ");
    };
    TestNameReporter.prototype.jasmineStarted = function (_suiteInfo) {
        this.stack = [];
        debug("stack: " + util_1.default.inspect(this.stack));
    };
    TestNameReporter.prototype.suiteStarted = function (result) {
        this.stack.push(result.description);
        debug("stack: " + util_1.default.inspect(this.stack));
    };
    TestNameReporter.prototype.specStarted = function (result) {
        this.stack.push(result.description);
        debug("stack: " + util_1.default.inspect(this.stack));
    };
    TestNameReporter.prototype.specDone = function (_result) {
        this.stack.pop();
        debug("stack: " + util_1.default.inspect(this.stack));
    };
    TestNameReporter.prototype.suiteDone = function (_result) {
        this.stack.pop();
        debug("stack: " + util_1.default.inspect(this.stack));
    };
    TestNameReporter.prototype.jasmineDone = function (_results) {
        this.stack = [];
    };
    return TestNameReporter;
}());
exports.default = TestNameReporter;
